### PR TITLE
GTEST/ARCH/TOOLS: Reworking wfe support

### PR DESCRIPTION
**Pull request by shamisp**
Fixing wait for event implementation and adding test to verify that WFE
does not introduce any substantial overheads.

Signed-off-by: Pavel Shamis (Pasha) <pasharesearch@gmail.com>

## What
Fixing WFE implementation and adding test to verify that it is correct one or at least does not hurt performance

## Why ?
Current implementation is broken.

## How ?
Adding test with and without WFE and check latency on small put messages.

### CODE REVIEW

**Comment by shamisp** [comment]
Ok to test

**Comment by shamisp** [comment]
@yosefe Do you know what they testing takes forever ?

**Comment by yosefe** [comment]
ok to test

**Comment by yosefe** [comment]
```
2020-11-02T17:49:51.2143730Z In file included from /scrap/azure/agent-05/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_wfe.cc:8:0:
2020-11-02T17:49:51.2146260Z /scrap/azure/agent-05/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.h: In instantiation of â€˜std::vector<ucp_test_param> enum_test_params(const string&) [with T = test_ucp_wfe; std::string = std::basic_string<char>]â€™:
2020-11-02T17:49:51.2148032Z /scrap/azure/agent-05/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_wfe.cc:97:1:   required from here
2020-11-02T17:49:51.2149318Z /scrap/azure/agent-05/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.h:329:27: error: â€˜get_test_variantsâ€™ is not a member of â€˜test_ucp_wfeâ€™
2020-11-02T17:49:51.2150185Z      T::get_test_variants(v);
2020-11-02T17:49:51.2150869Z                            ^
2020-11-02T17:49:51.2802859Z make[3]: *** [ucp/gtest-test_ucp_wfe.o] Error 1
2020-11-02T17:49:51.2804312Z make[3]: *** Waiting for unfinished jobs....
2020-11-02T17:49:52.5940131Z make[3]: Leaving directory `/scrap/azure/agent-05/AZP_WORKSPACE/2/s/build-test/test/gtest'
2020-11-02T17:49:52.5941266Z make[2]: *** [all-recursive] Error 1
2020-11-02T17:49:52.5942139Z make[2]: Leaving directory `/scrap/azure/agent-05/AZP_WORKSPACE/2/s/build-test/test/gtest'
2020-11-02T17:49:52.5942789Z make[1]: *** [all-recursive] Error 1
2020-11-02T17:49:52.5943494Z make[1]: Leaving directory `/scrap/azure/agent-05/AZP_WORKSPACE/2/s/build-test'
2020-11-02T17:49:52.5943909Z make: *** [all] Error 2
```

**Comment by shamisp** [comment]
@yosefe yep, saw this. I don't see this issue on by build machine but I will try to reproduce on some other machine.

**Comment by shamisp** [comment]
@yosefe the failure appears as a noise.

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    static ucs_log_func_rc_t
+    log_handler(const char *file, unsigned line, const char *function,
+                ucs_log_level_t level,
+                const ucs_log_component_config_t *comp_conf,
+                const char *message, va_list ap) {
+        // Ignore errors that transport cannot reach peer
+        if (level == UCS_LOG_LEVEL_ERROR) {
+            std::string err_str = format_message(message, ap);
+            if (strstr(err_str.c_str(), ucs_status_string(UCS_ERR_UNREACHABLE)) ||
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:40]:
```
if ((err_str.find(ucs_status_string(UCS_ERR_UNREACHABLE)) != std::string::npos) ||
    (err_str.find(ucs_status_string(UCS_ERR_UNSUPPORTED)) != std::string::npos)) {
}
```

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    std::stringstream ss;
+    ss << GetParam();
+    /* coverity[tainted_string_argument] */
+    ucs::scoped_setenv tls("UCX_TLS", "shm");
+    ucs::scoped_setenv warn_invalid("UCX_WARN_INVALID_CONFIG", "no");
+
+    test = tests[0];
+    if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_ARM_AARCH64) {
+	test.max *= UCP_ARM_PERF_TEST_MULTIPLIER;
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:83]: alignment, use 4 spaces

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	test.min /= UCP_ARM_PERF_TEST_MULTIPLIER;
+    }
+    /* Run ping-pong with no WFE and get latency values */
+    run_test(test, 0, false, "", "", &perf_value);
+    /* Run ping-pong with  WFE while re-using previous run numbers and a
+     * min/max boundary. The latency should stay nearly identical with 70
+     * percent margin */
+    test = tests[1];
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:91]: align by `=` for each line

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <gtest/common/test_perf.h>
+
+
+#define MB   pow(1024.0, -2)
+#define UCP_ARM_PERF_TEST_MULTIPLIER 2
+class test_ucp_wfe : public ucp_test, public test_perf {
+public:
+    static void get_test_variants(std::vector<ucp_test_variant>& variants) {
+    add_variant(variants, 0);
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:18]: alignment

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    int cache_perf_retry_count;
+    test_spec test;
+
+    std::stringstream ss;
+    ss << GetParam();
+    /* coverity[tainted_string_argument] */
+    ucs::scoped_setenv tls("UCX_TLS", "shm");
+    ucs::scoped_setenv warn_invalid("UCX_WARN_INVALID_CONFIG", "no");
+
+    test = tests[0];
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:81]: can somehow re-implement using `for()`?

**Review by dmitrygx — CHANGES_REQUESTED** [review]

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 5 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:81]: IMHO will make the code much less readable and self descriptive.

File: src/tools/perf/lib/libperf.c [hunk 6]
```diff
@@ -323,7 +323,8 @@ void ucx_perf_calc_result(ucx_perf_context_t *perf, ucx_perf_result_t *result)
     ucs_time_t median;
     double factor;
 
-    if (perf->params.test_type == UCX_PERF_TEST_TYPE_PINGPONG) {
+    if (perf->params.test_type == UCX_PERF_TEST_TYPE_PINGPONG ||
```
[the remark this answers is not in the document]
Comment (shamisp) [reply src/tools/perf/lib/libperf.c]: fixed

File: src/tools/perf/lib/ucp_tests.cc [hunk 7]
```diff
@@ -285,6 +286,13 @@ class ucp_perf_test_runner {
                     progress_responder();
                 }
                 return UCS_OK;
+	    case UCX_PERF_TEST_TYPE_PINGPONG_WFE:
```
[the remark this answers is not in the document]
Comment (shamisp) [reply src/tools/perf/lib/ucp_tests.cc]: fixed

File: src/tools/perf/lib/uct_tests.cc [hunk 8]
```diff
@@ -263,7 +263,8 @@ class uct_perf_test_runner {
                 return UCS_ERR_INVALID_PARAM;
             }
         case UCX_PERF_CMD_PUT:
-            if (TYPE == UCX_PERF_TEST_TYPE_PINGPONG) {
+            if (TYPE == UCX_PERF_TEST_TYPE_PINGPONG ||
```
[the remark this answers is not in the document]
Comment (shamisp) [reply src/tools/perf/lib/uct_tests.cc]: fixed

File: src/tools/perf/lib/ucp_tests.cc [hunk 9]
```diff
@@ -510,6 +518,7 @@ class ucp_perf_test_runner {
         /* coverity[switch_selector_expr_is_constant] */
         switch (TYPE) {
         case UCX_PERF_TEST_TYPE_PINGPONG:
+	case UCX_PERF_TEST_TYPE_PINGPONG_WFE:
```
[the remark this answers is not in the document]
Comment (shamisp) [reply src/tools/perf/lib/ucp_tests.cc]: fixed

File: src/ucs/arch/aarch64/cpu.h [hunk 10]
```diff
@@ -142,10 +142,11 @@ static inline void ucs_cpu_init()
 static inline void ucs_arch_wait_mem(void *address)
 {
     unsigned long tmp;
-    asm volatile ("ldxrb %w0, %1 \n"
+    asm volatile ("ldaxrb %w0, [%1] \n"
                   "wfe           \n"
                   : "=&r"(tmp)
-                  : "Q"(address));
+                  : "r"(address)
+		  : "memory");
```
[the remark this answers is not in the document]
Comment (shamisp) [reply src/ucs/arch/aarch64/cpu.h]: fixed

File: test/gtest/common/test_perf.cc [hunk 11]
```diff
@@ -253,7 +253,8 @@ test_perf::test_result test_perf::run_multi_threaded(const test_spec &test, unsi
 }
 
 void test_perf::run_test(const test_spec& test, unsigned flags, bool check_perf,
-                         const std::string &tl_name, const std::string &dev_name)
+                         const std::string &tl_name, const std::string &dev_name,
+                         double *perf_value)
```
[the remark this answers is not in the document]
Comment (shamisp) [reply test/gtest/common/test_perf.cc:257]: It is pointer to the value that by default is NULL. Since I cannot NULL reference, it has to be pointer.

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 1 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc]: fixed

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 2 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc]: fixed

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 3 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc]: fixed

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 4 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc]: fixed

**Comment by shamisp** [comment]
@dmitrygx - most the issues have been addressed

**Comment by shamisp** [comment]
@yosefe @dmitrygx - I cannot access "MellanoxLab New" failures.

**Comment by dmitrygx** [comment]
> @yosefe @dmitrygx - I cannot access "MellanoxLab New" failures.

related to #5337

bot:pipe:retest

File: test/gtest/common/test_perf.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (dmitrygx) [reply test/gtest/common/test_perf.cc:257]: > It is pointer to the value that by default is NULL. Since I cannot NULL reference, it has to be pointer.

I see

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 12]
```diff
@@ -0,0 +1,102 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+*
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:3]: ARM?

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 13]
```diff
@@ -0,0 +1,102 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:2]: 2020

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+*
+* Copyright (C) UT-Battelle, LLC. 2015. ALL RIGHTS RESERVED.
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:4]: not needed?

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+*/
+
+#include "ucp_test.h"
+
+#include <gtest/common/test_perf.h>
+
+
+#define MB   pow(1024.0, -2)
+#define UCP_ARM_PERF_TEST_MULTIPLIER 2
+class test_ucp_wfe : public ucp_test, public test_perf {
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:15]: add two spaces after macro declaration

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+public:
+    static void get_test_variants(std::vector<ucp_test_variant>& variants) {
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:17]: `std::vector<ucp_test_variant> &variants`

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    run_test(test, 0, false, "", "", &perf_value);
+    /* Run ping-pong with  WFE while re-using previous run numbers and a
+     * min/max boundary. The latency should stay nearly identical with 250
+     * percent margin. When WFE does not work as expected the slow down is
+     * typically 10x */
+    test     = tests[1];
+    test.max = perf_value * 2.5;
+    test.min = perf_value * 0.7;
+    cache_perf_retry_count = ucs::perf_retry_count;
+    ucs::perf_retry_count = 3; /* Have to be set to 1 otherwise performance
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:95]: align by `=`

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 5 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_wfe.cc:80]: maybe better to split into 2 different test cases: `put_latency_without_wfe` and `put_latency_with_wfe`?

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+UCS_TEST_P(test_ucp_wfe, envelope) {
+    double perf_value = 0;
+    int cache_perf_retry_count;
+    test_spec test;
+
+    std::stringstream ss;
+    ss << GetParam();
+    /* coverity[tainted_string_argument] */
+    ucs::scoped_setenv tls("UCX_TLS", "shm");
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:77]: it's not a good way
better to use
```
UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wfe, shm, "shm,")
```
instead of
```
UCP_INSTANTIATE_TEST_CASE(test_ucp_wfe)
```

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    ucs::scoped_setenv warn_invalid("UCX_WARN_INVALID_CONFIG", "no");
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:78]: this is not needed then

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+    test = tests[0];
+    if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_ARM_AARCH64) {
+        test.max *= UCP_ARM_PERF_TEST_MULTIPLIER;
+        test.min /= UCP_ARM_PERF_TEST_MULTIPLIER;
+    }
+    /* Run ping-pong with no WFE and get latency values */
+    run_test(test, 0, false, "", "", &perf_value);
+    /* Run ping-pong with  WFE while re-using previous run numbers and a
```
Comment (dmitrygx) [inline test/gtest/ucp/test_ucp_wfe.cc:87]: extra whitespace before `WFE`

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 12 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:3]: Yes, I have to add

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 13 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:2]: We only bump the date if the file was update by the vendor in this particular year.

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 14 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:4]: Since it was in the original perf file that I copied for this file, I have to preserve the copyright (BSD3 requirements)

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 16 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:17]: I don't think this aligns with the format that is used in other instances https://github.com/openucx/ucx/blob/f473cd28a1e3e7d691fbaa6d27acafa05396d444/test/gtest/ucp/test_ucp_wireup.cc#L24

Generally speaking I have no religion about this but if I change it will be inconsistent with other instances.

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 18 above)
Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:77]: @dmitrygx It is single use case intentionally. The test check that WFE implementation does not introduce substantial latency overheads. First we are running the test without WFE and then follow with test with WFE. Then we compare difference between the two runs and signal failure if it is above the threshold.

Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc:77]: I will change to UCP_INSTANTIATE_TEST_CASE_TLS

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 13 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_wfe.cc:2]: > We only bump the date if the file was update by the vendor in this particular year.

@shamisp thanks! I see

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 14 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_wfe.cc:3]: > Since it was in the original perf file that I copied for this file, I have to preserve the copyright (BSD3 requirements)

👍

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 16 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_wfe.cc:20]: not a problem, I think we need to clarify what's suitable in UCX

File: test/gtest/ucp/test_ucp_wfe.cc (hunk 18 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_wfe.cc]: > @dmitrygx It is single use case intentionally. The test check that WFE implementation does not introduce substantial latency overheads. First we are running the test without WFE and then follow with test with WFE. Then we compare difference between the two runs and signal failure if it is above the threshold.

I guess it is reply to https://github.com/openucx/ucx/pull/5861#discussion_r520691897
I see the idea, thank you.
Maybe define enum for `0` and `1` that are used to get a perf test suite

Comment (shamisp) [reply test/gtest/ucp/test_ucp_wfe.cc]: Fixed 2d1d360

**Comment by shamisp** [comment]
@dmitrygx - please let me know if there is anything else.

**Review by dmitrygx — APPROVED** [review]

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+                const char *message, va_list ap) {
+        // Ignore errors that transport cannot reach peer
+        if (level == UCS_LOG_LEVEL_ERROR) {
+            std::string err_str = format_message(message, ap);
+            if ((err_str.find(ucs_status_string(UCS_ERR_UNREACHABLE)) != std::string::npos) ||
+                (err_str.find(ucs_status_string(UCS_ERR_UNSUPPORTED)) != std::string::npos)) {
+                UCS_TEST_MESSAGE << err_str;
+                return UCS_LOG_FUNC_RC_STOP;
+            }
+        }
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:48]: why needed?

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    const static test_spec tests[];
+};
+
+
+enum {
+    UCX_PERF_TEST_LAT_NO_WFE,
+    UCX_PERF_TEST_LAT_WITH_WFE
+};
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:59]: can just define 2 test specs (without an array)

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    test.min               = perf_value * 0.7;
+    cache_perf_retry_count = ucs::perf_retry_count;
+    ucs::perf_retry_count  = 3; /* Have to be set to 1 otherwise the performance
+				                  measurement is ignored */
+    run_test(test, 0, true, "", "");
+    /* restore global value for perf retry */
+    ucs::perf_retry_count = cache_perf_retry_count;
+}
+
+UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wfe, shm, "shm,")
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:105]: maybe move all code from test_ucp_wfe.cc to test_ucp_perf.c ? can add the test case directoly to test_ucp_perf, no need to duplicate code

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+    /* Run ping-pong with WFE while re-using previous run numbers as a min/max
+     * boundary. The latency of the WFE run should stay nearly identical with 250
+     * percent margin. When WFE does not work as expected the slow down is
+     * typically 10x-100x */
+    test                   = tests[UCX_PERF_TEST_LAT_WITH_WFE];
+    test.max               = perf_value * 2.5;
+    test.min               = perf_value * 0.7;
+    cache_perf_retry_count = ucs::perf_retry_count;
+    ucs::perf_retry_count  = 3; /* Have to be set to 1 otherwise the performance
+				                  measurement is ignored */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:99]: indent

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    double perf_value = 0;
+    int cache_perf_retry_count;
+    test_spec test;
+
+    test = tests[UCX_PERF_TEST_LAT_NO_WFE];
+    if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_ARM_AARCH64) {
+        test.max *= UCP_ARM_PERF_TEST_MULTIPLIER;
+        test.min /= UCP_ARM_PERF_TEST_MULTIPLIER;
+    }
+    /* Run ping-pong with no WFE and get latency reference values */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:88]: pls add space line before

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+    run_test(test, 0, false, "", "", &perf_value);
+    /* Run ping-pong with WFE while re-using previous run numbers as a min/max
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:90]: pls add space line before

File: test/gtest/ucp/test_ucp_wfe.cc [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+     * boundary. The latency of the WFE run should stay nearly identical with 250
+     * percent margin. When WFE does not work as expected the slow down is
+     * typically 10x-100x */
+    test                   = tests[UCX_PERF_TEST_LAT_WITH_WFE];
+    test.max               = perf_value * 2.5;
+    test.min               = perf_value * 0.7;
+    cache_perf_retry_count = ucs::perf_retry_count;
+    ucs::perf_retry_count  = 3; /* Have to be set to 1 otherwise the performance
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wfe.cc:98]: hacky
let's pass retry_count to run_test as optional arg (which has default value)
